File: AspNetMvcUnleashed/0672329980_source/Chapter11/VB/MvcApplication1/MvcApplication1.Tests/Controllers/SimpleMovieControllerTests.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports System.Web.Mvc
Imports MvcFakes

<TestClass()> _
 Public Class SimpleMovieControllerTests

    <TestMethod()> _
    Public Sub IndexAddsMoviesToCache()
        ' Arrange
        Dim cache = New FakeCache()
        Dim service = New SimpleMovieService(New FakeSimpleMovieRepository(), cache)
        Dim controller = New SimpleMovieController(service)

        ' Act
        controller.Index()

        ' Assert
        Assert.IsInstanceOfType(cache("movies"), GetType(IEnumerable(Of Movie)))
    End Sub

    <TestMethod()> _
    Public Sub IndexRetrievesMovieFromCache()
        ' Arrange movies
        Dim movies = New List(Of Movie)()
        movies.Add(Movie.CreateMovie(1, "Star Wars", "Lucas", DateTime.Parse("1/1/1977")))

        ' Arrange cache
        Dim cache = New FakeCache()
        cache("movies") = movies

        ' Arrange controller
        Dim service = New SimpleMovieService(New FakeSimpleMovieRepository(), cache)
        Dim controller = New SimpleMovieController(service)

        ' Act
        Dim results = CType(controller.Index(), ViewResult)

        ' Assert
        Dim movieResults = CType(results.ViewData.Model, List(Of Movie))
        CollectionAssert.AreEqual(movies, movieResults)
    End Sub

End Class

File: AspNetMvcUnleashed/0672329980_source/Chapter11/CS/MvcApplication1/MvcApplication1.Tests/Controllers/SimpleMovieControllerTests.cs
using System;
using System.Collections.Generic;
using System.Web.Mvc;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using MvcApplication1.Controllers;
using MvcApplication1.Models;
using MvcApplication1.Tests.Fakes;
using MvcFakes;

namespace MvcApplication1.Tests.Controllers
{
    [TestClass]
    public class SimpleMovieControllerTests
    {

        [TestMethod]
        public void IndexAddsMoviesToCache()
        {
            // Arrange
            var cache = new FakeCache();
            var service = new SimpleMovieService(new FakeSimpleMovieRepository(), cache); 
            var controller = new SimpleMovieController(service);

            // Act
            controller.Index();

            // Assert
            Assert.IsInstanceOfType(cache["movies"], typeof(IEnumerable<Movie>));
        }


        [TestMethod]
        public void IndexRetrievesMovieFromCache()
        {
            // Arrange movies
            var movies = new List<Movie>();
            movies.Add(Movie.CreateMovie(1, "Star Wars", "Lucas", DateTime.Parse("1/1/1977")));

            // Arrange cache
            var cache = new FakeCache();
            cache["movies"] = movies;
            
            // Arrange controller
            var service = new SimpleMovieService(new FakeSimpleMovieRepository(), cache);
            var controller = new SimpleMovieController(service);

            // Act
            var results = (ViewResult)controller.Index();

            // Assert
            var movieResults = (List<Movie>)results.ViewData.Model;
            CollectionAssert.AreEqual(movies, movieResults);
        }



    }
}
